README.rst
CrossCompute Extensions for Jupyter Notebook
============================================

Thanks to `Salah Ahmed <https://github.com/salah93>`_ for prototyping these extensions.

Installation
------------
::

    sudo dnf -y install procps
    pip install crosscompute-notebook-extensions
    jupyter nbextension install \
        --py --sys-prefix crosscompute_notebook_extensions
    jupyter nbextension enable \
        --py --sys-prefix crosscompute_notebook_extensions
    jupyter serverextension enable \
        --py --sys-prefix crosscompute_notebook_extensions

Usage via command-line
----------------------
::

    git clone https://github.com/crosscompute/default-notebooks
    cd default-notebooks/Examples/Python
    crosscompute serve "Suggest Concepts.ipynb"

Usage via Jupyter Notebook
--------------------------
::

    cd tests/notebooks
    jupyter notebook

- Click the blue paper plane icon (or press SHIFT-C SHIFT-P) to preview the notebook as a tool.
- Click the red paper plane icon (or press SHIFT-C SHIFT-D) to deploy the notebook as a tool.


CHANGES.rst
0.5
---
- Add ToolDeployJson extension
- Add ConfigurationUpdateJson extension
- Show errors if deploy fails

0.4
---
- Preview only if first code cell has the word CrossCompute
- Show errors if preview fails
- Ignore IPython magic
- Improve unicode support
- Save tool template if first cell is markdown
- Save result template if last cell is markdown

0.3
---
- Support custom base_url

0.2
---
- Add example notebooks
- Add ToolPreview redirect
- Fix Python 3 support
- Replace pgrep with psutil to support non-POSIX systems

0.1
---
- Add ToolPreviewJson extension
- Extract notebook support from crosscompute package


tests/test_macros.py
from crosscompute_notebook_extensions.macros import get_tool_name


def test_get_tool_name():
    f = get_tool_name
    assert f('/tmp/Untitled.ipynb') == 'untitled'
    assert f('AddIntegers') == 'add-integers'
    assert f('CountWords2016') == 'count-words-2016'


crosscompute_notebook_extensions/ipynb.py
import nbconvert
from collections import OrderedDict
from copy import deepcopy
from crosscompute.configurations import find_tool_definition
from crosscompute.exceptions import CrossComputeError
from crosscompute.extensions import ToolExtension
from crosscompute.scripts import corral_arguments
from crosscompute.types import RESERVED_ARGUMENT_NAMES
from invisibleroads_macros.configuration import (
    make_absolute_paths, make_relative_paths)
from invisibleroads_macros.disk import (
    copy_folder, copy_text, make_enumerated_folder, HOME_FOLDER)
from invisibleroads_macros.text import unicode_safely
from jinja2 import Environment, PackageLoader
from nbconvert.exporters import PythonExporter
from os.path import abspath, dirname, join
from six import string_types

from .macros import get_tool_name, load_notebook


JINJA2_ENVIRONMENT = Environment(loader=PackageLoader(
    'crosscompute_notebook_extensions'))


class IPythonNotebookTool(ToolExtension):

    @classmethod
    def prepare_tool_definition(self, path, with_debugging=False):
        tool_name = get_tool_name(path)
        script_folder = make_enumerated_folder(join(
            HOME_FOLDER, '.crosscompute', tool_name, 'tools'))
        prepare_script_folder(script_folder, path, tool_name, with_debugging)
        return find_tool_definition(script_folder, default_tool_name=tool_name)


def prepare_script_folder(
        script_folder, notebook_path, tool_name, with_debugging=False):
    notebook = load_notebook(notebook_path)
    notebook_folder = dirname(abspath(notebook_path))
    tool_arguments = load_tool_arguments(notebook)
    tool_arguments = make_absolute_paths(tool_arguments, notebook_folder)
    try:
        tool_arguments = corral_arguments(script_folder, tool_arguments)
    except IOError as e:
        raise CrossComputeError({e.args[0]: 'path not found (%s)' % e.args[1]})
    kw = {'with_debugging': with_debugging}
    # Save notebook contents
    copy_folder(script_folder, notebook_folder)
    # Save script
    script_content = prepare_script(tool_arguments, notebook)
    script_name = 'run.py'
    copy_text(join(script_folder, script_name), script_content)
    # Save tool template
    cell = notebook['cells'][0]
    if cell['cell_type'] == u'markdown':
        template_name = 'tool.md'
        copy_text(join(script_folder, template_name), cell['source'])
        kw['tool_template_path'] = template_name
    # Save result template
    cell = notebook['cells'][-1]
    if cell['cell_type'] == u'markdown':
        template_name = 'result.md'
        copy_text(join(script_folder, template_name), cell['source'])
        kw['result_template_path'] = template_name
    # Save configuration
    command_name = notebook['metadata']['kernelspec']['name']
    configuration_content = prepare_configuration(
        tool_name, command_name, script_name, make_relative_paths(
            tool_arguments, script_folder), **kw)
    configuration_name = 'cc.ini'
    copy_text(join(script_folder, configuration_name), configuration_content)


def load_tool_arguments(notebook):
    g, l = OrderedDict(), OrderedDict()
    code_cells = get_code_cells(notebook)
    if not code_cells:
        raise CrossComputeError('cannot make a tool without code')
    try:
        exec(code_cells[0]['source'], g, l)
    except Exception as e:
        raise CrossComputeError(e)
    d = OrderedDict()
    for k, v in l.items():
        if k.startswith('__'):
            continue
        d[unicode_safely(k)] = unicode_safely(v)
    return d


def prepare_script(tool_arguments, notebook):
    script_lines = []
    script_lines.append('from sys import argv')
    for index, k in enumerate(tool_arguments):
        v = tool_arguments[k]
        if isinstance(v, string_types):
            line_template = '{argument_name} = argv[{i}]'
        else:
            line_template = '{argument_name} = {value_type}(argv[{i}])'
        script_lines.append(line_template.format(
            argument_name=k, i=index + 1, value_type=type(v).__name__))
    script_lines.extend(['\n'] + JINJA2_ENVIRONMENT.get_template(
        'python.jinja2').render().splitlines())
    notebook = deepcopy(notebook)
    code_cells = get_code_cells(notebook)
    code_cells[0]['source'] = '\n'.join(script_lines)
    script_content = nbconvert.export(PythonExporter, notebook)[0]
    return script_content


def get_code_cells(notebook):
    code_cells = []
    for cell in notebook['cells']:
        if cell['cell_type'] == u'code':
            code_cells.append(cell)
    return code_cells


def prepare_configuration(
        tool_name, command_name, script_name, tool_arguments,
        tool_template_path=None, result_template_path=None,
        with_debugging=False):
    configuration_lines = []
    configuration_lines.append('[crosscompute %s]' % tool_name)
    configuration_lines.append('command_template = %s %s %s' % (
        command_name, script_name, ' '.join(
            '{%s}' % x for x in tool_arguments)))
    for k, v in tool_arguments.items():
        if k in RESERVED_ARGUMENT_NAMES:
            continue
        configuration_lines.append('x.%s = %s' % (k, v))
    if tool_template_path:
        configuration_lines.append(
            'tool_template_path = %s' % tool_template_path)
    if result_template_path:
        configuration_lines.append(
            'result_template_path = %s' % result_template_path)
    if with_debugging:
        configuration_lines.append('show_raw_output = True')
    configuration_content = '\n'.join(configuration_lines)
    return configuration_content


crosscompute_notebook_extensions/assets/tools.js
var notebook;

define([
  'base/js/namespace',
  'base/js/dialog'
], function(jupyter, dialog) {

  var $modal;
  var $toolbar = $('#maintoolbar-container');
	var keyboard_manager = jupyter.keyboard_manager;
	notebook = jupyter.notebook;

  function preview_tool() {
    process_notebook('preview', 'preview');
  }

  function deploy_tool() {
    if (notebook.notebook_name == 'Untitled.ipynb') {
      return show_modal('Tool name required', 'Your notebook is Untitled. Please give a name to your notebook. The name of the notebook will become the name of the tool.');
    }
    process_notebook('deploy', 'deployment');
  }

  function process_notebook(verb, noun) {
    function render_text(x) {
      return x.replace(/X/g, noun);
    }

		var code_cell = notebook.container.find('.code_cell').first().data('cell');
		if (code_cell === undefined || !/crosscompute/i.test(code_cell.get_text())) {
      show_modal(render_text('Tool X cancelled'), '<p>This notebook does not appear to be a CrossCompute Tool.</p><p><a href="https://crosscompute.com/create#create-tools" target="_blank">Please make sure the first code cell contains the word CrossCompute</a>.</p>');
      return;
    }

    show_modal(render_text('Preparing tool X...'), 'Please be patient.');
    notebook.events.one('notebook_saved.Notebook', function() {
      $.ajax({
        method: 'post',
        url: notebook.base_url + 'extensions/crosscompute/X.json'.replace(/X/g, verb),
        data: {
          '_xsrf': get_cookie('_xsrf'),
          'notebook_path': notebook.notebook_path
        }
      }).fail(function(jqXHR) {
        switch(jqXHR.status) {
          case 400:
            var d = jqXHR.responseJSON;
            show_modal(render_text('Tool X failed'), d.text ? '<pre>' + d.text + '</pre>' : 'Unable to reach server.');
            break;
          case 401:
            show_modal('Server token required', '<textarea class="form-control"></textarea>');
            var $textarea = $modal.find('textarea');
            $modal.one('hidden.bs.modal', function() {
              var server_token = $.trim($textarea.val());
              if (!server_token.length) return;
              update_configuration('server_token', server_token);
            });
            break;
          case 403:
            show_modal('Account sign-in required', 'Anonymous sessions cannot X tools. <a href="https://crosscompute.com" target="_blank">Please sign in and start a new session</a> to X this tool.'.replace(/X/g, verb));
            break;
        }
      }).done(function(d) {
        var body = '<p><a href="X" target="_blank">Click here to access your tool</a>.</p>'.replace(/X/g, d.tool_url);
        if (verb == 'preview') {
          body += '<p>Note that your tool preview will stop working after the notebook session ends. Deploy the tool using the Red Button to make it available on <a href="https://crosscompute.com" target="_blank">CrossCompute</a>.</p>';
        } else if (verb == 'deploy') {
          body += '<p>Its visibility is <b>hidden</b>, which means that only users with the exact url will be able to access the tool. You can choose to make the tool public by clicking the eye icon at the top of the deployed tool. If you make it public, the tool will appear on your profile.</p>';
        }
        show_modal(render_text('Tool X succeeded'), body);
      });
    });
    notebook.save_notebook();
    setTimeout(function() {
      if ($('.modal').length > 1) {
        $modal.modal('hide');
      }
    }, 500);
  }

  function update_configuration(variable_name, variable_value) {
    $.ajax({
      method: 'post',
      url: notebook.base_url + 'extensions/crosscompute/configure.json',
      data: {
        '_xsrf': get_cookie('_xsrf'),
        'variable_name': variable_name,
        'variable_value': variable_value,
      },
      success: function(d) {
        show_modal('Configuration update succeeded', 'Please retry your request.');
      },
      error: function(jqXHR) {
        show_modal('Configuration update failed', 'Could not update configuration at this time.');
      }
    });
  }

  function get_cookie(name) {
    var r = document.cookie.match('\\b' + name + '=([^;]*)\\b');
    return r ? r[1] : undefined;
  }

  function show_modal(title, body) {
    if (!$modal) {
      $modal = dialog.modal({
        notebook: notebook,
        keyboard_manager: keyboard_manager,
        buttons: {'Close': {}}
      });
    }
    $modal.find('.modal-title').text(title);
    $modal.find('.modal-body').html(body);
    $modal.modal('show');
  }

  function set_toolbar_button_css(action_name, d) {
    var x = 'button[data-jupyter-action="X"]';
    $toolbar.find(x.replace('X', action_name)).css(d);
  }

  function load_ipython_extension() {
    var actions = keyboard_manager.actions;
    var shortcuts = keyboard_manager.command_shortcuts;
    var preview_tool_action_name = actions.register({
      'icon': 'fa-paper-plane-o',
      'help': 'preview tool',
      'help_index': 'crosscompute-preview',
      'handler': preview_tool
    }, 'tool-preview', 'crosscompute');
    var deploy_tool_action_name = actions.register({
      'icon': 'fa-paper-plane-o',
      'help': 'deploy tool',
      'help_index': 'crosscompute-deploy',
      'handler': deploy_tool
    }, 'tool-deploy', 'crosscompute');

    shortcuts.add_shortcut('Shift-C,Shift-P', preview_tool_action_name);
    shortcuts.add_shortcut('Shift-C,Shift-D', deploy_tool_action_name);

    jupyter.toolbar.add_buttons_group([
      preview_tool_action_name,
      deploy_tool_action_name
    ]);

    set_toolbar_button_css(preview_tool_action_name, {'background-color': '#286090', 'color': '#e6e6e6'});
    set_toolbar_button_css(deploy_tool_action_name, {'background-color': '#d9534f', 'color': '#e6e6e6'});
  }

  return {
    load_ipython_extension: load_ipython_extension
  };

});


crosscompute_notebook_extensions/__init__.py
"""Server extensions for Jupyter Notebook

Adapted from prototypes by [Salah Ahmed](https://github.com/salah93)
"""
import atexit
import requests
from configparser import ConfigParser
from crosscompute.exceptions import CrossComputeError
from invisibleroads_macros.disk import compress
from invisibleroads_macros.text import unicode_safely
from notebook.base.handlers import IPythonHandler
from notebook.utils import url_path_join
from os.path import expanduser, join
from psutil import process_iter
from requests.exceptions import ConnectionError
from signal import SIGINT
from subprocess import Popen, PIPE
from tempfile import gettempdir
from time import sleep
from tornado import web

from .ipynb import IPythonNotebookTool
from .settings import S


class ConfigurationUpdateJson(IPythonHandler):

    def post(self):
        variable_name = self.get_argument('variable_name')
        if variable_name not in ['server_token']:
            self.set_status(400)
            return self.write({variable_name: 'unexpected'})
        variable_value = self.get_argument('variable_value')
        configuration = ConfigParser()
        configuration_path = get_configuration_path()
        configuration.read(configuration_path)
        if 'crosscompute-website' not in configuration:
            configuration.add_section('crosscompute-website')
        configuration_section = configuration['crosscompute-website']
        configuration_section[variable_name] = variable_value
        configuration.write(open(configuration_path, 'wt'))
        S[variable_name] = variable_value
        self.write({})


class ToolPreviewJson(IPythonHandler):

    def post(self):
        stop_servers()
        notebook_path = self.get_argument('notebook_path')
        tool_port = S['tool_port']
        process_arguments = [
            'crosscompute', 'serve', notebook_path,
            '--host', S['tool_host'],
            '--port', str(tool_port),
            '--base_url', S['tool_base_url'],
            '--without_browser', '--without_logging', '--with_debugging']
        for x in 'brand_url', 'website_name', 'website_owner':
            y = expect_variable(x, '')
            if y:
                process_arguments.extend(('--' + x, y))
        open(join(gettempdir(), 'preview.sh'), 'wt').write(' '.join(
            process_arguments))
        process = Popen(process_arguments, stderr=PIPE)
        d = {}
        for x in range(10):
            try:
                requests.get('http://127.0.0.1:%s' % tool_port)
            except ConnectionError:
                sleep(1)
            else:
                status_code = 200
                d['tool_url'] = self._get_tool_url()
                break
            if process.poll():
                status_code = 400
                d['text'] = unicode_safely(process.stderr.read().strip())
                break
        else:
            status_code = 400
        self.set_status(status_code)
        self.write(d)

    def _get_tool_url(self):
        tool_base_url = S['tool_base_url']
        if tool_base_url == '/':
            request_host = self.request.host.split(':')[0]
            tool_port = S['tool_port']
            tool_url = '//%s:%s' % (request_host, tool_port)
        else:
            tool_url = tool_base_url
        return tool_url


class ToolDeployJson(IPythonHandler):

    def post(self):
        try:
            server_token = expect_variable('server_token')
        except KeyError as e:
            self.set_status(401)
            return self.write({})
        server_url = expect_variable('server_url', 'https://crosscompute.com')
        notebook_id = expect_variable('notebook_id', '')
        notebook_path = self.get_argument('notebook_path')
        try:
            tool_definition = IPythonNotebookTool.prepare_tool_definition(
                notebook_path, with_debugging=False)
        except CrossComputeError as e:
            self.set_status(400)
            return self.write({'text': str(e)})
        archive_path = compress(tool_definition['configuration_folder'])

        response = requests.post(server_url + '/tools.json', headers={
            'Authorization': 'Bearer ' + server_token,
        }, data={
            'notebook_id': notebook_id,
            'environment_level': expect_variable('environment_level', 0),
            'processor_level': expect_variable('environment_level', 0),
            'memory_level': expect_variable('environment_level', 0),
        } if notebook_id else {}, files={
            'tool_folder': open(archive_path, 'rb'),
        })
        status_code = response.status_code
        if status_code != 200:
            self.set_status(status_code)
            return self.write({})
        try:
            d = response.json()
            tool_url = d['tool_url']
        except Exception:
            self.set_status(503)
            return self.write({})
        self.write({'tool_url': server_url + tool_url})


def get_configuration_path():
    return expanduser('~/.crosscompute/.settings.ini')


def load_jupyter_server_extension(notebook_app):
    base_url = notebook_app.base_url
    configure_url = get_extension_url(base_url, 'configure')
    preview_url = get_extension_url(base_url, 'preview')
    deploy_url = get_extension_url(base_url, 'deploy')
    # Configure settings
    settings = notebook_app.config.get('CrossCompute', {})
    S['tool_host'] = settings.get('host', '127.0.0.1')
    S['tool_port'] = int(settings.get('port', 4444))
    S['tool_base_url'] = '/' if base_url == '/' else preview_url
    # Configure routes
    host_pattern = r'.*$'
    if notebook_app.password:
        for x in ConfigurationUpdateJson, ToolPreviewJson, ToolDeployJson:
            x.post = web.authenticated(x.post)
    web_app = notebook_app.web_app
    web_app.add_handlers(host_pattern, [
        (configure_url + '.json', ConfigurationUpdateJson),
        (preview_url + '.json', ToolPreviewJson),
        (deploy_url + '.json', ToolDeployJson),
    ])
    # Set exit hooks
    atexit.register(stop_servers)


def get_extension_url(base_url, verb):
    namespace_url = get_namespace_url(base_url)
    return url_path_join(namespace_url, verb)


def get_namespace_url(base_url):
    return url_path_join(base_url, '/extensions/crosscompute')


def stop_servers():
    for process in process_iter():
        if process.name() != 'crosscompute':
            continue
        x = process.cmdline()
        if len(x) > 3 and x[2] == 'serve':
            process.send_signal(SIGINT)


def expect_variable(variable_name, default_value=None):
    try:
        return S[variable_name]
    except KeyError:
        pass
    configuration = ConfigParser()
    configuration_path = get_configuration_path()
    configuration.read(configuration_path)
    try:
        value = configuration['crosscompute-website'][variable_name]
    except KeyError:
        pass
    else:
        S[variable_name] = value
        return value
    if default_value is None:
        raise KeyError
    return default_value


def _jupyter_nbextension_paths():
    return [{
        'section': 'notebook',
        'src': 'assets',
        'dest': 'crosscompute',
        'require': 'crosscompute/tools',
    }]


def _jupyter_server_extension_paths():
    return [{
        'module': 'crosscompute_notebook_extensions',
    }]
